source/framework/vx_distribution.c

```diff
@@ -18,13 +18,80 @@
 
 #include <vx_internal.h>
 
-vx_distribution VX_API_CALL vxCreateDistribution(
-    vx_context context, vx_size num_bins, vx_int32 offset, vx_uint32 range)
+static vx_distribution ownCreateDistribution(vx_reference scope, vx_size num_bins, vx_int32 offset, vx_uint32 range, vx_bool is_virtual);
+static vx_status isDistributionCopyable(vx_distribution input, vx_distribution output);
+static vx_status copyDistribution(vx_distribution input, vx_distribution output);
+static vx_status swapDistribution(vx_distribution input, vx_distribution output);
+static vx_status VX_CALLBACK distributionKernelCallback(vx_enum kernel_enum, vx_bool validate_only, vx_enum optimization, const vx_reference params[], vx_uint32 num_params);
+
+/*! \brief This function is called to find out if it is OK to copy the input to the output.
+ * Bins, offset & range must be the same
+ * \returns VX_SUCCESS if it is, otherwise another error code.
+ * 
+ */
+static vx_status isDistributionCopyable(vx_distribution input, vx_distribution output)
+{
+    if ((vx_enum)vx_true_e == tivxIsReferenceMetaFormatEqual((vx_reference)input, (vx_reference)output))
+    {
+         return VX_SUCCESS;
+    }
+    else
+    {
+        return VX_ERROR_NOT_COMPATIBLE;
+    }
+}
+
+/*! \brief Copy input to output
+ * The input must be copyable to the output; checks done already.
+ * Note that locking a reference actually locks the context, so we only lock
+ * one reference!
+
+ */
+static vx_status copyDistribution(vx_distribution input, vx_distribution output)
+{
+    return (ownCopyReferenceGeneric((vx_reference)input, (vx_reference)output));
+}
+
+/*! \brief swap input and output pointers
+ * Input and output must be swappable; checks done already.
+ */
+static vx_status swapDistribution(vx_distribution input, vx_distribution output)
+{
+    return ownSwapReferenceGeneric((vx_reference)input, (vx_reference)output);
+}
+
+/* Call back function that handles the copy, swap and move kernels */
+static vx_status VX_CALLBACK distributionKernelCallback(vx_enum kernel_enum, vx_bool validate_only, vx_enum optimization, const vx_reference params[], vx_uint32 num_params)
+{
+    /*
+        Decode the kernel operation - simple version!
+    */
+    vx_distribution input = (vx_distribution)params[0];
+    vx_distribution output = (vx_distribution)params[1];
+    switch (kernel_enum)
+    {
+        case VX_KERNEL_COPY:    return validate_only ? isDistributionCopyable(input, output) : copyDistribution(input, output);
```

Comment:
Can we go a step further in these cases and remove the static functions: isDistributionCopyable, swapDistribution, and copyDistribution by doing:

 case VX_KERNEL_COPY:    return validate_only ? tivxIsReferenceMetaFormatEqual(input, output) : ownCopyReferenceGeneric((vx_reference)input, (vx_reference)output);

Same true of all other types which only need to call these generic function and need nothing beyond.

Reply:
And just to give you a hint to where I am going next, it we can do the above, then the next step is that this whole callback is now generic, can be renamed as genericReferenceKernelCallback and move to the reference.c.  Then the this genericReferenceKernelCallback function can be registered as part of the vxCreateDistribution function instead.  This would minimize most of these generic data object file changes.  Do you agree?

Reply:
I'll have a look and do also here some generic functions

Reply:
done for:
vx_array : done ony isArrayCopyable cannot be currently generic
vx_matrix: done only isMatrixCopyable cannot be currently generic
vx_lut: done
scalar: not needed
vx_tensor: done only for copyTensor
vx_convolution: done for the copy and swap functions
vx_distrbution: done
vx_pyramid: not applicable
vx_image / raw_image : not applicable
vx_objarray: not applicable